### ISSUE TITLE
[servers] porting audio servers for Emscripten/WebAssembly

### ISSUE THREAD

**Issue by brianlheim** [supercollider/supercollider#5224]
## Motivation

Prior discussion on mailing list here: https://www.listarc.bham.ac.uk/lists/sc-dev/msg00227.html

> I know main devs were not particularly keen on spending time on an
> scsynth wasm target when I asked last time, but perhaps there are other
> people out there who would be interested in exploring that avenue?
> 
> My personal interest got renewed after coming back to Scala.js for my
> client, and seeing that the JS back-end compiler these days is rock
> solid. I think it would be fantastic to make sound installations or
> sketches work in the client-side browser. I even hacked up a WebAudio
> UGen for my FScape non-realtime DSP system to test the thing -- see
> https://www.sciss.de/temp/soundprocesses.js/ -- but of course this was
> never designed for this scenario, so first of all, that's all JavaScript
> not Assembly, and it has a rather complicated streaming framework behind
> (Akka Stream) instead of the relatively simple scsynth graph processing;
> so no wonder it is not very efficient in CPU terms.
> 
> I believe the sweet spot is to run the client through JS (that could be
> ScalaCollider/SoundProcesses, supercollider.js, or other languages that
> probably can run in the browser, like Python) and the server through
> wasm. Csound people tell me wasm support was quite a good addition.
> 
> Now, as I've stated repeatedly, my C skills are very small, same for
> LLVM and emscripten, and I don't know how entangled the build targets
> for scsynth vs sclang are. But if anyone sees this as an opportunity to
> experiment, I'm happy to try to learn the necessary stuff to start
> building a branch of scsynth against wasm (if it's doable at all -- but
> why not?).
> 
> Feel free to reply to the thread or ping me directly.

- @Sciss 

## Description of Proposed Feature

this thread is mostly for discussing requirements and resources for the implementation.

**Comment by brianlheim** [comment]
some resources i found:

It looks like Emscripten uses OpenAL as their audio backend: https://emscripten.org/docs/porting/Audio.html

Here's an OpenAL audio API programmer's guide: https://www.openal.org/documentation/OpenAL_Programmers_Guide.pdf

here's info on porting: https://emscripten.org/docs/porting/index.html. some interesting limitations i noted:
- only non-blocking networking calls are allowed
- can't use x86 inline assembly
- virtual filesystem with some limitations due to sandboxing
- function pointers must be cast to the exact type (i think sclang may actually make use of this)
- exceptions turned off at optimized levels by default since they disable many optimizations

so far i'm seeing two requirements for a performant port of scsynth -- (1) provide an OpenAL backend and (2) aim to make it exception-free, with (2) as a stretch goal. both of those would take some work but i think they're very doable. also, there will probably need to be a lot of changes to the build system, i'm happy to give some pointers on that.

**Comment by Sciss** [comment]
Thanks Brian - I just noticed the ticket now, somehow I had skipped the e-mail. So here are the links from @Spacechild1 and others:

- [Csound](https://github.com/csound/csound) [link examples](https://github.com/kunstmusik/webaudio-csound-samples-example) [link Web IDE](https://github.com/csound/web-ide)
- [Pure Data](https://mathr.co.uk/empd/) (Claude Heiland-Allen)
- [Purr Data](https://cdm.link/2020/10/pure-data-is-running-in-the-browser-for-free-web-based-sound-patching/) (Albert Graef et al.) [link demo](http://cuinjune-purr-data.glitch.me/) [link repo](https://github.com/agraef/purr-data)

I think the common idea here is to run the DSP engine through wasm, and have the UI in JavaScript, thus allowing high level languages with GC.

**Comment by brianlheim** [comment]
> I think the common idea here is to run the DSP engine through wasm, and have the UI in JavaScript, thus allowing high level languages with GC.

sure, that makes sense to me. are you or @Spacechild1 interested in working on this?

**Comment by Sciss** [comment]
I'm happy to work on it, but as said - I'm really no C expert, so I'm not sure how much I can contribute in terms of actual code. For example, I wouldn't know what do say about the bullet points you listed, or OpenAL.

Naturally, I'm more interested in getting the server part done, as there are already a number of clients that could probably run on JS with little effort.

We could discuss this in the dev meeting. Probably one would create a branch and try step by step to establish a build script for emscripten, to get to a minimally working example (say: compile to wasm, be able to start the server and process an OSC message).

**Comment by Spacechild1** [comment]
I'm currently busy with personal projects. However, I'm planning to port one of my projects to WASM in the near future, and once I gained some experience I can help.

**Comment by brianlheim** [comment]
@Sciss @Spacechild1 sounds good! there's no rush on this.

**Comment by brianlheim** [comment]
note: may need to look into threading (is multithreading available in all browsers?)

**Comment by miguel-negrao** [comment]
It appears that it is: [SharedArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer#Browser_compatibility) and [Pthreads support](https://emscripten.org/docs/porting/pthreads.html)

**Comment by brianlheim** [comment]
reading further, there are many caveats listed on the pthread support article. i would not expect perfect compatibility with our codebase.

**Comment by Sciss** [comment]
@miguel-negrao would you be interested to try to work with me on such a project?

**Comment by miguel-negrao** [comment]
Sorry, closed this by mistake, not intentional. @sciss, sorry, I'm afraid not, but looks like interesting project.

**Comment by claudeha** [comment]
In my experiments, I've used Emscripten's SDL2 port for audio, it's callback driven (I only used it for output, not input or duplex). My experiments with Pure-data just combined libpd  (with some patches, the main one being the function pointer cast thing) compiled with Emscripten with Emscripten's SDL2 port.

In one project ("barry") I found that enabling Emscripten's "proxy to worker" build option was essential for making spiky workloads triggered by browser events (like recompiling the DSP chain when live coding) not interrupt the audio.  I think that meant I had to fix total memory limit at compile time, too (no memory growth), at least that's mentioned in the Makefile (but Emscripten is constantly improving, maybe that is fixed by now):
https://code.mathr.co.uk/barry/blob/6a2425b2f2127564b4c7058772e262aec15cd908:/Makefile#l85

**Comment by Sciss** [comment]
I think there are two steps with memory usage in SC - pre-allocated real-time memory, e.g. `DelayN` UGen, versus dynamically allocated memory, e.g `Buffer` with `BufDelayN`. Most other things are pre-allocated, like wire buffers, max nodes, max synth defs, etc. We could limit to pre-allocated only in the first step.

Another question I have - if one was to implement a minimum server with compatible API from "scratch"; would that work with the existing UGen plugins? I guess that's what Supernova did, using the existing plugins but implementing the server anew? Is anyone knowledgeable if the plugins could, in theory, be loaded also from a different C-like language, like Rust?

**Comment by Spacechild1** [comment]
SuperCollider's own plugins only depend on the public API, just like 3rd party plugins. They don't know about any Server internals.

> I guess that's what Supernova did, using the existing plugins but implementing the server anew?

Supernova also had to extend the plugin API, e.g. adding spinlocks to `World` and `SndBuf`. All Supernova plugins have to be compiled against this extended plugin API (i.e. with `SUPERNOVA` defined).

> Is anyone knowledgeable if the plugins could, in theory, be loaded also from a different C-like language, like Rust?

It is indeed possible to implement an alternative Server and load existing plugins, as long as the plugin API type and version match.

**Comment by xinaesthete** [comment]
I'd be certainly interested in using this, not sure if I'm based place for working on implementation but hope I can make some sort of contribution.

FWIW We use scsynth in an otherwise web-browser based environment for mutatorvr.co.uk... would be great to loose the dependency on external executable / proxy server etc.  Hopefully in the near future I'll extricate the SuperCollider related code from that into an open-source library (and make more cool stuff with it).

**Comment by Sciss** [comment]
So my first naive approach was try to run emscripten and see what happens:

```
$ emcmake cmake -DSC_WII=no -DSC_EL=no -DSUPERNOVA=no -DCMAKE_BUILD_TYPE="Release" --target scsynth ..
configure: cmake -DSC_WII=no -DSC_EL=no -DSUPERNOVA=no -DCMAKE_BUILD_TYPE=Release --target scsynth .. -DCMAKE_TOOLCHAIN_FILE=/home/hhrutz/Documents/devel/emsdk/upstream/emscripten/cmake/Modules/Platform/Emscripten.cmake -DCMAKE_CROSSCOMPILING_EMULATOR="/home/hhrutz/Documents/devel/emsdk/node/12.18.1_64bit/bin/node"
-- SuperCollider Version: 3.11.2
-- Building from branch develop, commit hash is 392d97dff
-- Compiling with Qt GUI
-- Could NOT find Pthreads (missing: PTHREADS_LIBRARY PTHREADS_INCLUDE_DIR) 
CMake Error at CMakeLists.txt:253 (message):
  cannot find libpthreads


-- Using bundled boost
CMake Error at external_libraries/CMakeLists.txt:24 (message):
  we need to implement boost_thread


-- Using bundled yaml-cpp
-- Could NOT find Pthreads (missing: PTHREADS_LIBRARY PTHREADS_INCLUDE_DIR) 
CMake Error at external_libraries/hidapi/CMakeLists.txt:32 (message):
  cannot find libpthreads


-- HIDAPI components:
--     hidapi_parser
CMake Error at external_libraries/hidapi/hidapi_parser/CMakeLists.txt:14 (target_link_libraries):
  Cannot specify link libraries for target "hidapi" which is not built by
  this project.


CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
PTHREADS_INCLUDE_DIR (ADVANCED)
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/tools
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi_parser
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi_parser
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi_parser
   used as include directory in directory /home/hhrutz/Documents/devel/supercollider/external_libraries/hidapi/hidapi_parser

-- Configuring incomplete, errors occurred!
See also "/home/hhrutz/Documents/devel/supercollider/build/CMakeFiles/CMakeOutput.log".
```
What would be my next steps? Would I remove the unsupported libraries from `CMakeLists.txt` and try again?

**Comment by Sciss** [comment]
My first attemps are here: https://github.com/Sciss/supercollider/tree/wasm

So far working to make the cmake configure complete

**Comment by Sciss** [comment]
With some more empty preprocessor ifdefs, I now get this list of undefined symbols:

 - `_Z17SC_NewAudioDriverP5World`
 - `_Z19initializeSchedulerv`
 - `_ZN13SC_Filesystem24defaultResourceDirectoryEv`
 - `_ZN13SC_Filesystem24defaultUserHomeDirectoryEv`
 - `_ZN13SC_Filesystem26defaultUserConfigDirectoryEv`
 - `_ZN13SC_Filesystem30defaultUserAppSupportDirectoryEv`
 - `_ZN13SC_Filesystem30isNonHostPlatformDirectoryNameERKNSt3__212basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEE`
 - `_ZN13SC_Filesystem32defaultSystemAppSupportDirectoryEv`
 - `_ZN13SC_Filesystem8freeGlobEPNS_4GlobE`
 - `_ZN13SC_Filesystem8globNextEPNS_4GlobE`
 - `_ZN13SC_Filesystem8makeGlobEPKc`
 - `oscTimeNow`
 - `pthread_setschedparam`
 - `sched_get_priority_max`
 - `sched_get_priority_min`
 - `server_timeseed`
 - `shmat`
 - `shmctl`
 - `shmdt`

I don't know if these are _all_ the undefined symbols, or just till the compiler gives up. Next step would be to find out how these symbols are generated; it looks like they are synthesized from some source symbols, like `_Z17SC_NewAudioDriverP5World` was probably just `SC_NewAudioDriverP5World` or ``SC_NewAudioDriver...World`. For this we'd need an appropriate `#include` in `SC_CoreAudio.h` that implements the OpenAL bindings. The file system stuff we can probably just put dummy values for now.  The last items, `shmctl` etc., seem to related to memory control, and perhaps are not found because `__LINUX__` is not defined, so just missing an additional condition.

If anyone wants to jump on right now, that would be great.

**Comment by Sciss** [comment]
Here are my thoughts now of how to go about adding the missing audio driver:

There are currently three options for Audio API on wasm: plain Web Audio API, OpenAL which has a back-end provided by emscripten, and SDL2, which apparently has a port that works with emscripten.
Currently the API string is `openal`, but it would have to be changed to `webaudio` or `sdl2` if we decide on one of these. Since Jack and PortAudio already cover a variety of platforms, I see
no big benefit in going through an extra indirection with OpenAL or SDL2 -- these are basically options if you want to port an existing C application to wasm, but since we have to write the
driver from scratch, it seems natural to skip this indirection and go for Web Audio API direct support instead.

The emscripten docs already show how to use the `EM_ASM` macros to invoke JavaScript calls from C. This would probably be the glue needed to set up the audio worklet processor.

Similar to `SC_PortAudio.cpp` and `SC_Jack.cpp`, we would add another file `SC_WebAudio.cpp` to `server/scsynth/` and register it in `server/scsynth/CMakeList.txt`.
In this file, the sub-class of `SC_AudioDriver` will be implemented.

**Comment by Sciss** [comment]
I added a dummy `SC_WebAudio.cpp`, and so it's possible to `make scsynth` (compiles and links), thus the next steps would be to understand how to download and run scsynth in the browser, I guess we could try to see if `scsynth -v` would work.

**Comment by claudeha** [comment]
On 23/12/2020 12:50, Hanns Holger Rutz wrote:
>
> we could try to see if |scsynth -v| would work.
>

https://code.mathr.co.uk/rodney/blob/9190f54d7e8b60329776ab6727190e5adf8d12d8:/index.html#l73 
is one project of mine that passes URL parameters to the emscripten 
runtime as command line arguments to main().

https://code.mathr.co.uk/dr1/blob/c080dd9cd5e5f5230d3638f6b020f6cd4ea80549:/harmonic-protocol/live/index.html#l147 
has basic interaction, with ccall into the main C program.
https://code.mathr.co.uk/dr1/blob/c080dd9cd5e5f5230d3638f6b020f6cd4ea80549:/harmonic-protocol/Makefile 
has some flags to make that work.

https://code.mathr.co.uk/barry/blob/6a2425b2f2127564b4c7058772e262aec15cd908:/web/index.html#l100 
passes a string to the C program, using --proxy-to-worker Emscripten 
flag to avoid audio xruns when the input() C functions takes too long:
https://code.mathr.co.uk/barry/blob/6a2425b2f2127564b4c7058772e262aec15cd908:/Makefile#l84
and EMSCRIPTEN_KEEPALIVE to prevent functions being optimized away:
https://code.mathr.co.uk/barry/blob/6a2425b2f2127564b4c7058772e262aec15cd908:/barry/barry.c#l362

**Comment by Sciss** [comment]
@claudeha thanks; do you have these programs – rodney, barry – running somewhere on a public server? I'm curious to see what they are doing. I also saw that you can use `--pre-js` to prepare, for example, `Module` with command line args.

**Comment by claudeha** [comment]
On 23/12/2020 13:45, Hanns Holger Rutz wrote:
>
> @claudeha <https://github.com/claudeha> thanks; do you have these 
> programs – rodney, barry – running somewhere on a public server?
>
https://mathr.co.uk/rodney/latest/ very cpu-heavy fractal graphics (no 
audio)
https://mathr.co.uk/barry/v2/ postfix (forth-like) stateless audio 
interpreter
https://mathr.co.uk/harmonic-protocol/ feedback drones with interactive 
control

**Comment by Sciss** [comment]
When actually starting scsynth, I'm running into some boost assertion errors; I also had to suppress some linker warnings related to shared-memory. My interpretation now is that while I add pthreads (and thereby shared-memory) options to the scsynth target, this is not true for the other targets it depends on, including boost. So in `server/scsynth/CMakeList.txt` I add

```
if (EMSCRIPTEN)
	set_target_properties(scsynth PROPERTIES LINK_FLAGS 
		"-s USE_PTHREADS=1 -s WASM=1")
endif()
```
is there a way to add these properties to all targets? How do I list all targets on which scsynth depends?

**Comment by Sciss** [comment]
Or perhaps - I have literally no knowledge of cmake - it's that the other thing that is linked is `libscsynth`? The linker errors were for `SC_ComPort.cpp`. I can see the following stuff being linked:

```
[ 20%] Linking CXX static library libboost_filesystem_lib.a
[ 25%] Linking C static library libtlsf.a
[ 29%] Linking CXX static library libboost_system_lib.a
[ 91%] Linking CXX static library libscsynth.a
[100%] Linking CXX executable scsynth.html
```
I don't know if that means any of this is linked with `wasm-ld`, if that only applies to scsynth? I guess everything that goes to `wasm-ld` must have the above linker properties.

**Comment by Sciss** [comment]
Nevermind, the boost problem is gone now that I set the linker flags for target `libscsynth` as well.

**Comment by Sciss** [comment]
OSC is working now, audio most likely; I cannot test it yet, because the plugins are not installed. I figured I should compile with `-DSTATIC_PLUGINS`, but this makes the linker choke:

```
em++   -stdlib=libc++ -g  -stdlib=libc++ -s USE_PTHREADS=1 -s WASM=1 -pthread --bind -s DISABLE_EXCEPTION_CATCHING=0 -s INITIAL_MEMORY=33554432 -s EXTRA_EXPORTED_RUNTIME_METHODS=["ccall","callMain"] --pre-js ../wasm/pre.js @CMakeFiles/scsynth.dir/objects1.rsp  -o scsynth.html @CMakeFiles/scsynth.dir/linklibs.rsp
error: undefined symbol: _Z10Delay_LoadP14InterfaceTable (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: __Z10Delay_LoadP14InterfaceTable may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _Z10Grain_LoadP14InterfaceTable (referenced by top-level compiled C/C++ code)
warning: __Z10Grain_LoadP14InterfaceTable may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _Z10Noise_LoadP14InterfaceTable (referenced by top-level compiled C/C++ code)
warning: __Z10Noise_LoadP14InterfaceTable may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _Z11Demand_LoadP14InterfaceTable (referenced by top-level compiled C/C++ code)
warning: __Z11Demand_LoadP14InterfaceTable may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _Z11DiskIO_LoadP14InterfaceTable (referenced by top-level compiled C/C++ code)
...
```
What's the correct way to build with static plugins? Do I need to add anything to the cmake files?

**Comment by Sciss** [comment]
Ok, if I add the corresponding UGen sources to scsynth target, it works.

Wow, it's playing sound! :boom:

**Comment by brianlheim** [comment]
STATIC_PLUGINS isn't supported, i was planning to remove this actually. do you absolutely need this to be a single binary?

**Comment by Sciss** [comment]
Yes, static linking is more or less required: https://github.com/emscripten-core/emscripten/wiki/Linking

But it _was working_ when adding the plugin source files. The necessary `#ifdef`s are there in the sources already. The result is now in my `wasm` branch, with demo running here: https://sciss.de/temp/scsynth.wasm/ (needs Chromium because I couldn't figure out yet how to set COOP/COEP http headers).

**Comment by brianlheim** [comment]
> static linking is more or less required: https://github.com/emscripten-core/emscripten/wiki/Linking

that's not quite my takeaway from a quick glance, if you can walk me through how you reached that conclusion i'd appreciate it. also just to be clear we are using dlopen for plugins, not dynamic linking. if we go the static linking route it makes it more complicated to integrate other plugins such as those in sc3-plugins.

do you know what this part means? maybe this is why you say it is impossible:

> The difference is that you do not use Module.dynamicLibraries; instead, you must load the side module into the filesystem, so that dlopen (or fopen, etc.) can access it

> But it was working when adding the plugin source files. The necessary #ifdefs are there in the sources already. 

that isn't the same thing as supporting a particular build configuration... we can start supporting it if we really need to and there's no significant drawback to keep it. anyway it's good that you got something building.

**Comment by Sciss** [comment]
well, yes, `dlopen` is easier than "generic" dynamic linking; basically you have to adjust the libraries (each plugin) so that they also emit .wasm (I think), and then these files have to be uploaded and put into a list known to emscripten, so they will be downloaded into an in-memory virtual file system, so that `dlopen` can find them. In short, much more complicated (from what I can see), and also slower, than using a static linking. I think for the time being and the projects considered for web integration, this is totally fine and easier to do. If someone wants to include third party plugins, well they have to add them to the cmakelist. Not a big problem IMHO.

**Comment by brianlheim** [comment]
> If someone wants to include third party plugins, well they have to add them to the cmakelist. Not a big problem IMHO.

it's okay if you don't feel up to doing more research/work, and i'm very grateful for what you've already done. however, that's not what i would call feasible for long-term support, so i'd acknowledge this as potential future work for anyone who's interested.

**Comment by brianlheim** [comment]
> The result is now in my wasm branch, with demo running here: https://sciss.de/temp/scsynth.wasm/ 

just tested it out with the d_bubbles and s_bubbles functions, this is so freaking cool, thank you!!!

your branch looks pretty good too, i have a few suggestions if you want them prior to making a PR, but also feel free to keep working on it if there's more you need to do. my first thought is that we could aim to put this in the 3.13 release, since 3.12 is likely coming up soon and we already have a major PR inbound for it (Bela support) plus other high priority things that need to be fixed.

**Comment by Sciss** [comment]
Thanks; yes I would like to make a PR soon, as when the new year kicks off, I will be dividing into a lot of other stuff. On my own list, this would be the most pressing items for the initial submission:

- [ ] copy audio input (respect number of inputs/outputs specified by user)
- [ ] emit proper OSC /reply messages
- [ ] try option for periods/buffer on the Web Audio side, similar to jack, to minimise drop outs
- [ ] run the clang-formatter

I can spend maximum one or two days on this, otherwise it would have to wait.

**Comment by Sciss** [comment]
A question on memory handling. My understanding of `small_scpacket` is that this is allocated on the stack, and so sending out reply messages, I would need to copy the packet onto a heap allocated memory block. Then, is `..._reply_func` in `SC_ComPort.cpp` always run on the same thread? Or could it be called from different threads? If from a single thread, I assume it would suffice to statically allocate the reply message once. Correct?

**Comment by brianlheim** [comment]
> I can spend maximum one or two days on this, otherwise it would have to wait.

let's wait then, there's no rush and the changes i would want you to make will probably take longer than that. once the Bela PR is merged i would be able to help a bit with the more complex C++ changes.

**Comment by brianlheim** [comment]
oh, or if you want you could submit a smaller "preparatory" PR with a few small and unintrusive things, like the change to .gitignore and the changes to get STATIC_PLUGINS working. but i would expect that to still take more than just a day or two to review and merge.

**Comment by Spacechild1** [comment]
@Sciss

> A question on memory handling. My understanding of small_scpacket is that this is allocated on the stack, and so sending out reply messages, I would need to copy the packet onto a heap allocated memory block.

Maybe I've missed something, but why can't the packet be on the stack?

> Then, is ..._reply_func in SC_ComPort.cpp always run on the same thread?

Scsynth: yes, Supernova: no.

Supernova actuallly has two non-realtime threads: the `system` thread (used for stage2 of async commands), and the `io` thread (used for trigger messages, node replies and for stage4 of async commands, i.e. `/done` messages). This means reply messages are mostly sent from the `io` thread - with one important exception: when stage2 of an async command fails, a `/fail` message is sent from the `system` thread! (Search for `report_failure` in `sc_osc_handler.cpp`)

**Comment by Sciss** [comment]
> Maybe I've missed something, but why can't the packet be on the stack?

That's when asynchronously calling JS. If I call synchronously, I end up in a JS worker thread which is almost certainly not what the JS client can use, also the global OSC endpoints are not visible. Thus, if I call asynchronously into the main JS thread, the stack will have become invalid by the time it is accessed.

Anyway, now I'm using a blocking call which I think is the cleanest and simplest solution; so I am able to pass the stack pointer.

> Scsynth: yes, Supernova: no.

Ok, good to know. For now, I'm just using scsynth, but solution should now work in both cases.

**Comment by Sciss** [comment]
I implemented the audio input now. I ran into a strange bug that I suspect is a bug in the implementation of `NumOutputBuses`. When I use this UGen I get an error

```
Uncaught RuntimeError: function signature mismatch
    at Graph_FirstCalc(Graph*) (http://127.0.0.1:8000/scsynth.wasm:wasm-function[2686]:0x70b98)
    at Group_Calc (http://127.0.0.1:8000/scsynth.wasm:wasm-function[2367]:0x59ef2)
    at World_Run(World*) (http://127.0.0.1:8000/scsynth.wasm:wasm-function[3855]:0xb821f)
    at SC_WebAudioDriver::WaRun() (http://127.0.0.1:8000/scsynth.wasm:wasm-function[11158]:0x43f0df)
    at emscripten::internal::MethodInvoker<void (SC_WebAudioDriver::*)(), void, SC_WebAudioDriver*>::invoke(void (SC_WebAudioDriver::* const&)(), SC_WebAudioDriver*) (http://127.0.0.1:8000/scsynth.wasm:wasm-function[11185]:0x440979)
    at SC_WebAudioDriver$WaRun [as WaRun] (eval at new_ (http://127.0.0.1:8000/scsynth.js:7673:27), <anonymous>:8:1)
    at ScriptProcessorNode.ad.proc.onaudioprocess (http://127.0.0.1:8000/scsynth.js:1699:1707)
```
which I read up correctly, is essentially a segfault caught within wasm. If I replace, in the pitch tracker example that is now inside `index.html` the expression `SoundIn.ar([0, 1])` by `In.ar(2, 2)` (assuming two output channels), the problem goes away.

I have no other clue than suspecting that that UGen somehow makes wrong assumptions, for example could it be a 32-bit problem? Is anyone successfully using that UGen in a 32-bit architecture (do we actually still have a 32-bit architecture that is used)?

Anyway; so I think once I understand how to run clang-format, I'm ready to make the PR.

**Comment by brianlheim** [comment]
32-bit SC should not be a problem unless there is platform specific code. we have been building and releasing 32-bit builds of SC on Windows for a long time

**Comment by Sciss** [comment]
Ok, well maybe it's accessing some data that wasn't initialised by `SC_WebAudio`; should be easy to find out with println debugging

**Comment by Sciss** [comment]
Seems to affect all `DefineInfoUnit`s, like `SampleRate` as well.

**Comment by Sciss** [comment]
The bug was a dangling extraneous Ctor parameter `int inNumSamples`. If I change the likes of

```cpp
void SampleDur_Ctor(Unit* unit, int inNumSamples);

void SampleDur_Ctor(Unit* unit, int inNumSamples) { ZOUT0(0) = unit->mWorld->mFullRate.mSampleDur; }
```
to
```cpp
void SampleDur_Ctor(Unit* unit);

void SampleDur_Ctor(Unit* unit) { ZOUT0(0) = unit->mWorld->mFullRate.mSampleDur; }
```
the crash goes away.

**Comment by Sciss** [comment]
I have untracked files

    external_libraries/portaudio_sc_org/

despite running `git submodule update --recursive`. Can you tell me what that is? Should I remove that directory?

**Comment by Sciss** [comment]
Is there an issue with using clang-format v9 ?

```
*** ERROR:
clang-format found, but incorrect version at /usr/bin/clang-format-9 with version: clang-format version 9.0.1-+20191211082622+c1a0a213378-1~exp1~20191211193200.102 

Accepted versions: 8.y.z
```
It's really a PITA getting LLVM to install on Debian stable. The oldest version supported seems clang 9. Would the formatting make any difference if I used 9?

**Comment by brianlheim** [comment]
yes, it would make a difference if you used 9, probably not major differences though. i see a clang 8 package in buster backports: https://packages.debian.org/buster-backports/devel/clang-format-8, can you install that?

it's not relevant what emscripten uses.

does this answer your git question? https://stackoverflow.com/questions/7993413/git-submodules-with-modified-and-untracked-content-why-and-how-to-remove-it#7993703

**Comment by xinaesthete** [comment]
I hadn't actually noticed that there was a demo webpage online to visit when I started writing this, although actually it doesn't quite seem to be working for me here: 
```
(index):190 wasm streaming compile failed: TypeError: Failed to execute 'compile' on 'WebAssembly': Incorrect response MIME type. Expected 'application/wasm'.
(index):190 falling back to ArrayBuffer instantiation
```
`Module.oscDriver` is then undefined, presumably having failed earlier steps it doesn't get as far as establishing that (although I don't think there was any error message in the HTML, only the console), although...

I have got it running locally, with the same above errors about incorrect MIME type / falling back to ArrayBuffer instantiation. On Chrome I do still get to hear pretty analog bubbles.

I just built this on mac os FWIW, with emscripten installed via homebrew. 
**&lt;noise&gt;**
I had a minor and likely unrelated snag in that something was a bit wrong with the brew/python setup on my machine (excuse the noise on this thread, maybe it'll help someone). I got this output:
```
Scanning dependencies of target boost_filesystem_lib
[  3%] Building CXX object external_libraries/CMakeFiles/boost_filesystem_lib.dir/boost/libs/filesystem/src/codecvt_error_category.cpp.o
/Users/peter/.pyenv/shims/python3: line 21: /usr/local/Cellar/pyenv/1.2.16/libexec/pyenv: No such file or directory
```
I could see that I had a 1.2.21 and no no 1.2.16, so as a quick hack I edited the file `/Users/peter/.pyenv/shims/python3` to point to the other folder and that allowed the build to work for now. I should sort out my python environment better - my next snag was that it was then complaining about `no module named SimpleHTTPServer` so I used `npx http-server` instead. I'll sort out my python environment when I'm dead (or have renewed interest in using it).
**&lt;/noise&gt;**

Loading the result in Firefox, I got an exception `SharedArrayBuffer is not defined` which I take it is because it requires cross-origin isolation, probably not too hard to set up but may benefit from documentation. Indeed it seems that this may be a limiting factor in getting it to work reliably across browsers? Am I right in thinking that SharedArrayBuffer is to do with the buffers used for the audio, whereas the other error about `falling back to ArrayBuffer instantiation` is to do with wasm compilation and unrelated?

In Safari it fails earlier due to lack of Atomics.

Anyway, it seems that supporting all browsers - particularly mobile - is likely out of scope at least for the time being. For now, I've only witnessed it working in Chrome, but I'm impressed that it went as smoothly as it did for me.

**Comment by Sciss** [comment]
Hi Peter. I believe the MIME error is harmless, that's why it says it's falling back to a different approach of loading the file. For example, I have a copy running here where the MIME type is correct: https://scsynth.iem.sh/

`Module.oscDriver` is only defined when the main wasm is executed, so until you press 'Boot', that symbol is indeed undefined. One could initialise it to an empty object if needed.

I run on Linux, so I used the default Emsdk installation described there. FWIF, installing Python or Ruby correctly is always a PITA, especially since they come in various versions and flavours. As you write, you can use any other web server (I have nginx installed as well, for example).

The Firefox error is described in the `README_WASM` with further links. Unfortunately `SharedArrayBuffer` is needed as long scsynth is multi-threaded (with more work, it might be possible to compile it single threaded?), and Firefox is more restrictive here thanks to the Spectre vulnerability; and I have no idea how to set up the COOP/COEP for my personal website (I don't have full Apache access). Kind of silly that it doesn't even work on localhost. So it's Chrome/Chromium for now, as you have witnessed. I hope we can find a solution for Firefox. Perhaps we find a web host with correct COOP/COEP settings that can be used for some pieces

**Comment by Sciss** [comment]
@xinaesthete P.S. if you want me to include your macOS build instructions, please send them (the successful path :)

**Comment by xinaesthete** [comment]
Ah yes @Sciss I'm afraid it did indeed slip my mind to press the button to boot the server, thanks for your response.

Wasn't too worried about the MIME thing, but thanks for clarification. I think I partly included that info just as a result of not sufficiently editing after realising it wasn't source of actual failure to work (and indeed perhaps I could have benefitted from more thorough parsing of the readme).

As far as MacOS instructions go, I think the only small data point is that if people prefer to `brew install emscripten` rather than installing it via other means then it seems to work absolutely fine - notwithstanding that there's always the possiblity of something else in the setup not being quite right. I don't think I'd encourage anyone to make the same hacky change I did if they find themselves with the specific problem that I had, but there's a slim chance that someone with a similar issue (probably not doing anything to do with SuperCollider) might stumble on that info via their search engine of choice and find it vaguely useful (so, although it's really just noise here, maybe it serves some purpose that it exists somewhere findable, but I certainly wouldn't suggest polluting the actual docs with it).

Given that homebrew itself seems to not be doing a particularly good job of keeping my env in order in this case, I'm not sure I'd go out of my way to encourage use of that, either, but of course many of us prefer to use some sort of package management rather than have random build tools scattered about their system, all things being equal.

Another small point is that it might be worth changing the .gitignore - I have 652 locally added files as a result of running the build. Actually, there may even be something to said for allowing the actual wasm output into git but I'm certainly not strongly arguing it should be.

I was vaguely wondering earlier if there might be a case for a docker container, but it seems fairly unncessary. I might try building on Windows at some point and see if it works as easily as it did on mac.

**Comment by Sciss** [comment]
> I have 652 locally added files as a result of running the build.

That's strange - I have none. What are the paths?

**Comment by xinaesthete** [comment]
> That's strange - I have none. What are the paths?

There are quite a few directly in `wasm`, and theres also `wasm/CMakeFiles/`, `wasm/external_libraries/`, `wasm/lang/`, `wasm/platform`, `wasm/server/` etc... actually come to think of it this could well be a result of not fully following some step in setting up the emsdk environment after installing with homebrew, or some other brew-related vaguery, I'll try to look into that. My only prior use of emscripten was in another project where they had set it up under a docker container (which was what had primed me to think about that earlier).

**Comment by Sciss** [comment]
There shouldn't be created _any_ files in the `wasm` directory, the build system doesn't even know about it. Merely the `make.sh` copies the compiled `.js` and `.wasm` files there. Are you sure you are running the build (as usual) from within the `./build` directory? Like

    $ rm -r build
    $ mkdir build
    $ cd build
    $ ../wasm/build.sh

?

**Comment by xinaesthete** [comment]
Ah, please excuse my inattentiveness, yes indeed of course that was it 🤦

**Comment by Sciss** [comment]
Somebody just told me they have it working in Firefox 85 (I was running myself with 78), so that's some good sign. There's another demo with ScalaCollider as client: https://www.sciss.de/temp/scalacollider.js/

**Comment by joshpar** [comment]
I just tried with Safari and that throws an exception - what would effect which browsers can use it?

Thanks

Josh

**Comment by Sciss** [comment]
@joshpar I think this: https://caniuse.com/sharedarraybuffer

**Comment by adadu** [comment]
https://www.sciss.de/temp/scalacollider.js/ worked for me on Firefox 84.0.1 on Arch Linux - if that is useful

**Comment by Spacechild1** [comment]
https://www.sciss.de/temp/scalacollider.js/ also works for me on Firefox 83 on Windows 7.

Awesome work!

**Comment by rd--** [comment]
Hello @Sciss 

This is amazing work!

I'm listening to the 'walters' example as I type this, it works perfectly!

And here `top` says chromium is at 20%CPU, with is pretty much exactly what it says of scsynth with the same graph.

I'm super impressed!

This will be very useful!
 
I see you wrote above (in Dec 2020) that you'd "like to make a PR soon", so I'm wondering:

1.Are you still working on and/or using this?

2. Would it be possible to get it merged into scsynth proper?

3. Are there instructions somewhere on how to set this up locally?

4. Is https://github.com/Sciss/supercollider/tree/wasm still where the work on this lives?

I know, so far, nothing at all about web audio, but I'm happy to poke around if that's the best way for now!

Kind regards
Rohan

**Comment by Sciss** [comment]
Hi Rohan, thanks for your message. There was a PR - #5309 - I guess it was still a bit WIP - it got closed now, because there is a new issue of rebasing it and bringing it up to date; you'll find it here: #5571 

I think it's viable as it is, needs a few things still missing (notably reporting CPU usage in /status.reply), and the long term goal would be to have it run in a dedicated web worker, so there is less chance of drop outs due to cooperative multitasking with JavaScript.

**Comment by rd--** [comment]
Hello @Sciss,

Thanks very kindly.  I've finally got it working here, it's very nice, it's so simple to work with!

Also, I realised I didn't need to actually compile this, I could just save the files from the addresses above, they're completely portable!

One question, moderate size osc messages fail with a "oscDriver.send: message size exceeded" warning.  Is the size limit a variable somewhere?

Best,
Rohan

**Comment by Sciss** [comment]
Hi @rd-- ; the failure indicates you are sending a message larger than 64K:

- https://github.com/Sciss/supercollider/blob/9f3d9e5315e554ab01f35fdb2c18a072984ed9b0/server/scsynth/SC_ComPort.cpp#L155
- https://github.com/Sciss/supercollider/blob/9f3d9e5315e554ab01f35fdb2c18a072984ed9b0/server/scsynth/SC_ComPort.cpp#L515

You could try to increase that, but I think 64K is also the hard limit for UDP, and I have never seen a situation where scsynth would run with OSC messages larger than 64K, so perhaps you think the message size is moderate, but actually you are somehow producing a much larger message than intended?

**Comment by rd--** [comment]
Hello @Sciss,

Thanks for the pointers!

The messages are simple d_recv sends.

Ordinarily they'd get rewritten as d_load, but I'm not sure if that can work in this context?

Or is there a sort of pseudo file-system we can use? 

I'm still not setup to actually compile this here, but I will in the next week, I'll try adjusting the limits then and report back.

Thanks again.

**Comment by Sciss** [comment]
Ah, yes, that's a problem. No there is no virtual file system yet that you could use for d_load. You'd have to see what happens when you increase the size in the source code beyond 64K ; OTTOMH, I don't know if this causes allocation errors or not.

**Comment by rd--** [comment]
Hello @Sciss,

I set kTextBufSize to 524288 and compiled scsynth.wasm and it works perfectly!

This is brilliant, thanks again!

I guess the next thing I'd like to do is make a version that has the sc3-plugins builtin.

Has someone done this?  Are there instructions?  I'm very inexpert I'm afraid.

Best,
Rohan

**Comment by Sciss** [comment]
As far as I remember, the standard plugins are directly included in the sources list of the build and thus statically included. I think the easiest to create a build with sc3-plugins is to do something similar, i.e. to make sure their sources are directly included in the sources list, rather than running two builds (I have no experience with library loading in wasm)

**Comment by Sciss** [comment]
This commit might help: https://github.com/Sciss/supercollider/commit/6002639143bd64c273b614309633fccaaf2e79d7

There is a flag `STATIC_PLUGINS` set, which is used in `SC_Lib_Cintf.cpp` ; I think you could add further source directories in `server/scsynth/CMakeLists.txt` and then initialise them in `SC_Lib_Cintf.cpp`.

**Comment by rd--** [comment]
Hello again,

Yes, that works perfectly, thanks kindly!

I've only added a few Ugens so far but once I have a moderate collection
put together I'll put it up somewhere.

Another question, FBSineC from the core plugins seems to crash the web
assembly scsynth here.

It prints "uncaught exception: 14739752"

Do you know how I would go about working out why?

I see above that you tracked done some errors in the standard UGens
earlier, is it possible to get a stack trace?

Best,
Rohan

**Comment by claudeha** [comment]
On 09/02/2022 07:20, rd-- wrote:
> It prints "uncaught exception: 14739752"

Could be that C++ exception catching is disabled by default in Emscripten?

https://emscripten.org/docs/porting/exceptions.html

**Comment by rd--** [comment]
Thanks.

From what I've read I think it should be set, by DISABLE_EXCEPTION_CATCHING in the clause below?

https://github.com/dylans/supercollider/blob/64cabdb0a913814d57b2f14d8f0e86ada4154162/server/scsynth/CMakeLists.txt#L298

I'm not entirely sure if this gets applied to plugins though, or only to scsynth.

So far all of the Ugens that crash in Wasm but not in the ordinary build are from:

https://github.com/supercollider/supercollider/blob/develop/server/plugins/ChaosUGens.cpp

And none I've tried of those work.  Perhaps a clue...

Also perhaps it'd be easier to run some kind of validator on scsynth, valgrind or such, I'm not sure.

Anyway, no real news yet.

**Comment by iamjonny** [comment]
Hi, I've just found this whilst reviewing the status of web-based supercollider.

The demo boots and works 👍
Android 12; CPH2307 Build/RKQ1.211103.002
Chrome 112.0.5615.47

Is this possible/planned/on the roadmap, and if so any ETA or links to project and status (and some help info for noobs!).

Thank you 👍

**Comment by dataexcess** [comment]
Hello all,

I was wondering if the supercollider port would also be possible without webworkers / multithreading. For the artpiece I want to make, it will be embedded on another site in an iframe. However the embedding isn't working since the iframe only has 'allow-scripts' as sandbox arguments. And so the communication between the client and the sc server is not possible. (allow-same-origin would fix this) but I do not control the code of the third party website that embeds my website. 

I guess it's not possible since supercollider inherintly has a client and server architecture? Or maybe earlier versions of SC would make this possible? e.g. SC 2. 

Thnx

**Comment by the-awb** [comment]
Hello, rooting around the interwebs and came across this project. Really great to see a supercollider browser port! Well done folks.

Fired up this demo link at all booted fine, getting bubble sounds gallore. https://www.sciss.de/temp/scalacollider.js/

I have some silly obvious questions...

Is there any sort of guide to using it? 
How could one go about hooking up an instance of sclang to the browser's server? 
Where is the repo? 
How could I go about building a web project around it (I'm familiar with node, react etc)?

Thanks, and again congrats on the great work!

**Comment by dataexcess** [comment]
> Hello, rooting around the interwebs and came across this project. Really great to see a supercollider browser port! Well done folks.
> 
> Fired up this demo link at all booted fine, getting bubble sounds gallore. https://www.sciss.de/temp/scalacollider.js/
> 
> I have some silly obvious questions...
> 
> Is there any sort of guide to using it? How could one go about hooking up an instance of sclang to the browser's server? Where is the repo? How could I go about building a web project around it (I'm familiar with node, react etc)?
> 
> Thanks, and again congrats on the great work!

You can't. Currently there's no sclang support, it's only server-side. You have to use server messages as explained in the WASM readme file. You can build your own Synthdef in sclang and then instead of the "(...).add" use "().asBytes.printOn(Post); nil", this will print an array of bytes which you can use to build your synthdef by sending messages to the wasm webworker. you first boot it, then you send a message to build the synthdef on the server, then send a message to play it on the server. 

So no way of using any sclang constructs/features for now. If you're planning on using patterns or anything like that, it might be good to consider the server-side alternatives such as the Deman ugens etc... or you could of course write your own javascript around the osc messages to make your own higher level constructs ;)

**Comment by the-awb** [comment]
Great, the WASM readme was really want I needed. 

Thanks!

Does anyone have any project they'd like to share for reference? I'd be especially interested to see if someone has combined this with the supercollider.js node package.

**Comment by JordanHendersonMusic** [comment]
@capital-G can this be closed?
